conans/model/requires.py

```diff
@@ -220,12 +220,13 @@ def __eq__(self, other):
         propagating includes or libs or run info or both are visible or the reference is the same,
         we consider the requires equal, so they can conflict"""
         return (self.ref.name == other.ref.name and self.build == other.build and
+                self.options == other.options and
```

Comment:
This will most likely produce lots of issues down the road. Having different options will produce different branches, with different dependencies to the same package, but with different options. But the C/C++ compile and link model doesn't support it in the general case, it requires a lot of encapsulation (shared library linking a static one as pure implementation detail, no API or ABI leaking), that doesn't happen in most cases.

Reply:
Thanks for the feedback @memsharded - I am beginning to think its more a consequence of how we integrate conan into our build system where we may build many different binaries per conanfile. In this case there is one binary which wants a shared zlib and a different binary that wants a static libpng and consequently static zlib, but they are both getting their dependencies from the same conanfile.

I think the main concern about this though was just the fact that this felt unexpected/"footgun"-ish because the conflict happened with no warning, it seemed strange that libpng could silently and implicitly (using a default shared=False, not explicitly stated) override the latter requires for zlib which had explicitly stated shared=True.

Reply:
This is a result of many factors, mainly:
- Conan graph is expanded depth-first. This is an important factor to enable the correct propagation and computation of things like version ranges, conflicts, and specially the propagation of requirements-traits (influenced by package-type)
- Evaluating the Conan graph is costly. For every node in the dependency graph it is necessary to download it from the server if not cached, unzip it, load the recipe, and evaluate it. That means that any graph computation that requires back-tracking could be like one order of magnitude slower than what it is today, and that would make it unusable for many users.
- If a node can use their settings and options to define conditional dependencies (and they often do), only the first set of options values used in the first expansion of a node can be used as valid, because these options condition all the behavior of the package and the dependency subgraph of that package.
- If we wanted to re-evaluate the options when a diamond in the dependency graph is closed, it would be necessary to invalidate all the upstream subgraphs, but also to invalidate and re-compute all the downstream graph, because a given package being shared or static also determines everything downstream, and that could result in further invalidations of the graph downstream. And that would basically require a lot of backtracking.

Conan 1 implemented something like that and it was just very fragile and problematic, not to say adding a ton of complexity that was blocking the possibility of adding many of the new features that Conan 2 has already released.

So as a summary:
- Defining options in recipes is not a strong definition, but a low priority suggestion.
- If recipes want to block and raise when the dependencies doesn't define some options values, they can implement a ``validate()`` method
- In general, the recommended way to define ``options`` that are to be enforced in the graph is defining them in the ``profile`` files
- Lastly, if some package is going to always be used with a given ``option``, then, removing that option would be the recommended way. If the issue is packages from ConanCenter, working from a fork as recommended in https://docs.conan.io/2/devops/using_conancenter.html would allow to easily implement these changes.

I hope that this clarifies a bit the rationale of this behavior.

Reply:
Thanks, I really appreciate you taking the time to explain the reasoning for that.

Reply:
Thanks for your feedback! 

We really thought about this while designing Conan 2, and we really tried to come up with something feasible to be able to manage options more thoroughly, but the trade-offs were too costly, so we finally decided to go with the current behavior and promote a more global definition of options. With this approach it seems things are going pretty smoothly, as it also helps with other challenging aspects of the graph resolution as the usage of lockfiles.